include/exec/finally.hpp

```diff
@@ -0,0 +1,371 @@
+
+    template <class _Sigs>
+    using __result_type =
+      __if<std::is_same<__result_type_<_Sigs>, __not_a_variant>, __ignore, __result_type_<_Sigs>>;
+
+    template <class _ResultType, class _ReceiverId>
+    struct __final_operation_base {
+      using _Receiver = __t<_ReceiverId>;
+
+      _Receiver __receiver_{};
+      __manual_lifetime<_ResultType> __result_{};
+    };
+
+    namespace __rec {
+      template <class _Sig>
+      struct __rcvr_vfun;
+
+      template <class _Sigs>
+      struct __vtable {
+        struct __t;
+      };
+
+      template <class _Sigs, class _Env>
+      struct __ref;
+
+      template <class _Tag, class... _As>
+      struct __rcvr_vfun<_Tag(_As...)> {
+        void (*__fn_)(void*, _As...) noexcept;
+      };
+
+      template <class _Rcvr>
+      struct __rcvr_vfun_fn {
+        template <class _Tag, class... _As>
+        constexpr void (*operator()(_Tag (*)(_As...)) const noexcept)(void*, _As...) noexcept {
+          return +[](void* __rcvr, _As... __as) noexcept -> void {
+            _Tag{}((_Rcvr&&) *(_Rcvr*) __rcvr, (_As&&) __as...);
+          };
+        }
+      };
+
+      template <class... _Sigs>
+      struct __vtable<completion_signatures<_Sigs...>> {
+        struct __t : __rcvr_vfun<_Sigs>... {
+          template <class _Rcvr>
+            requires receiver_of<_Rcvr, completion_signatures<_Sigs...>>
+          friend const __t*
+            tag_invoke(__any::__create_vtable_t, __mtype<__t>, __mtype<_Rcvr>) noexcept {
+            static const __t __vtable_{{__rcvr_vfun_fn<_Rcvr>{}((_Sigs*) nullptr)}...};
+            return &__vtable_;
+          }
+        };
+      };
+
+      template <class... _Sigs, class _Env>
+      struct __ref<completion_signatures<_Sigs...>, _Env> {
+        using __vtable_t = stdexec::__t<__vtable<completion_signatures<_Sigs...>>>;
+
+        class __t {
+          void* __receiver_;
+          _Env __env_;
+          const __vtable_t* __vtable_;
+
+          template <std::same_as<__t> _Self>
+          friend _Env tag_invoke(get_env_t, const _Self& __self) noexcept {
+            return __self.__env_;
+          }
+
+          template <__completion_tag _Tag, __decays_to<__t> _Self, class... _As>
+            requires __one_of<_Tag(_As...), _Sigs...>
+          friend void tag_invoke(_Tag, _Self&& __self, _As&&... __as) noexcept {
+            (*static_cast<const __rcvr_vfun<_Tag(_As...)>*>(__self.__vtable_)->__fn_)(
+              __self.__receiver_, (_As&&) __as...);
+          }
+
+         public:
+          template <__none_of<__t, const __t> _Rcvr>
+            requires receiver_of<_Rcvr, completion_signatures<_Sigs...>>
+          __t(_Rcvr& __rcvr) noexcept
+            : __receiver_{std::addressof(__rcvr)}
+            , __env_{get_env(__rcvr)}
+            , __vtable_{__any::__create_vtable(__mtype<__vtable_t>{}, __mtype<_Rcvr>{})} {
+          }
+        };
+      };
```

Comment:
I take it this type erasure is necessary because you want to test the nothrow-ness of `connect` before the receiver type is known. The type-erasure is going to lead to worse runtime performance. I don't know whether that's the right trade-off or not.

Reply:
Exact. I did it to start a discussion about exactly this. It wouldn't be needed if completion signatures would depend on the receiver and not its environment. Also, an operation is only fully described if a receiver is known. Who knows what else limits the true transparency of signatures